src/lib/reasoners/bitv_rel.ml

```diff
@@ -171,45 +173,67 @@ end = struct
     let changed = if !changed then SX.add r t.changed else t.changed in
     { changed; bitlists }
 
-  let get r t =
+  let update_signed ex { Bitv.value; negated } t bl =
+    let bl = if negated then Bitlist.lognot bl else bl in
+    update_leaf ex value t bl
+
+  let update ex r t bl =
+    fst @@ List.fold_left (fun (t, bl) { Bitv.bv; sz } ->
+        (* Extract the bitlist associated with the current component *)
+        let mid = Bitlist.width bl - sz in
+        let bl_tail =
+          if mid = 0 then Bitlist.empty else
+            Bitlist.extract bl 0 (mid - 1)
+        in
+        let bl = Bitlist.extract bl mid (Bitlist.width bl - 1) in
+
+        match bv with
+        | Bitv.Cte z ->
+          (* Nothing to update, but still check for consistency! *)
+          ignore @@ Bitlist.intersect bl (Bitlist.exact sz z Ex.empty) ex;
+          t, bl_tail
+        | Other r -> update_signed ex r t bl, bl_tail
+        | Ext (r, r_size, i, j) ->
+          (* r<i, j> = bl -> r = ?^(r_size - j - 1) @ bl @ ?^i *)
```

Comment:
Yes, I kind abuse notations and use `bl` to denote an element of the domain here. I think it's a bit easier to think in terms of equality rather than set inclusion but I can change the `=` to `in`.